sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureSeekableByteChannel.java

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.blob.nio;
+
+import com.azure.core.util.logging.ClientLogger;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.NonReadableChannelException;
+import java.nio.channels.NonWritableChannelException;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileAttribute;
+
+/**
+ * A byte channel that maintains a current position.
+ * <p>
+ * This type is primarily offered to support some jdk convenience methods such as
+ * {@link Files#createFile(Path, FileAttribute[])} which requires opening a channel and closing it. A channel may only
+ * be opened in read mode OR write mode. It may not be opened in read/write mode. Seeking is supported for reads, but
+ * not for writes. Modifications to existing files is not permitted--only creating new files or overwriting existing
+ * files.
+ * <p>
+ * This type is not threadsafe to prevent having to hold locks across network calls.
+ * <p>
+ * {@link NioBlobInputStream} and {@link NioBlobOutputStream} are the preferred types for reading and writing blob data
+ * and are used internally by this type.
+ */
+public class AzureSeekableByteChannel implements SeekableByteChannel {
+    private final ClientLogger logger = new ClientLogger(AzureSeekableByteChannel.class);
+
+    private final NioBlobInputStream reader;
+    private final NioBlobOutputStream writer;
+    private long position;
+    private boolean closed = false;
+    private final Path path;
+
+    AzureSeekableByteChannel(NioBlobInputStream inputStream, Path path) {
+        this.reader = inputStream;
+        inputStream.mark(Integer.MAX_VALUE);
+        this.writer = null;
+        this.position = 0;
+        this.path = path;
+    }
+
+    AzureSeekableByteChannel(NioBlobOutputStream outputStream, Path path) {
+        this.writer = outputStream;
+        this.reader = null;
+        this.position = 0;
+        this.path = path;
+    }
+
+    @Override
+    public int read(ByteBuffer dst) throws IOException {
+        AzurePath.ensureFileSystemOpen(this.path);
+        validateOpen();
+        validateReadMode();
+
+        // See comments in position()
+        if (this.position > this.size()) {
+            return -1;
+        }
+
+        int count = 0;
+
+        int len = dst.remaining();
+        byte[] buf = new byte[len];
```

Comment:
A small performance improvement would be checking `ByteBuffer.hasArray()` and if so use the backing array instead of instantiating a new one.

```java
byte[] buf;
if (dst.hasArray()) {
    buf = dst.array();
} else {
    buf = new byte[len];
}

// Reading logic

if (!dst.hasArray()) {
    dst.put(buf, 0, count);
}
```

Reply:
Good idea!